src/query_graph/graph_path.rs

```diff
@@ -238,4 +481,770 @@ impl OpGraphPath {
+        let mut product = Vec::with_capacity(num_options);
+
+        // Compute the cartesian product.
+        for _ in 0..num_options {
+            let num_simultaneous_paths = options_for_each_path
+                .iter()
+                .zip(option_indexes.iter())
+                .fold(0, |acc, (options, option_index)| {
+                    acc + options[*option_index].0.len()
+                });
+            let mut simultaneous_paths = Vec::with_capacity(num_simultaneous_paths);
+
+            for (options, option_index) in options_for_each_path.iter().zip(option_indexes.iter()) {
+                simultaneous_paths.extend(options[*option_index].0.iter().cloned());
+            }
+            product.push(SimultaneousPaths(simultaneous_paths));
+
+            for (options, option_index) in
+                options_for_each_path.iter().zip(option_indexes.iter_mut())
+            {
+                if *option_index == options.len() - 1 {
+                    *option_index = 0
+                } else {
+                    *option_index += 1;
+                    break;
+                }
+            }
+        }
+
+        product
+    }
+
+    /// Given 2 `SimultaneousPaths` that represent 2 different options to reach the same query leaf
+    /// field, checks if one can be shown to be always "better" (more efficient/optimal) than the
+    /// other one, regardless of any surrounding context (i.e. regardless of what the rest of the
+    /// query plan would be for any other query leaf field).
+    ///
+    /// Note that this method is used on the final options of a given "query path", so all the
+    /// heuristics done within `GraphPath` to avoid unnecessary options have already been applied
+    /// (e.g. avoiding the consideration of paths that do 2 successive key jumps when there is a
+    /// 1-jump equivalent), so this focus on what can be done is based on the fact that the path
+    /// considered is "finished".
+    ///
+    /// Returns the comparison of the complexity of `self` and `other` (e.g. `Ordering::Less` means
+    /// `self` is better/has less complexity than `other`). If we can't guarantee anything (at least
+    /// "out of context"), then we return `Ordering::Equal`.
+    fn compare_options_complexity_out_of_context(
+        &self,
+        other: &SimultaneousPaths,
+    ) -> Result<Ordering, FederationError> {
+        match (self.0.len() == 1, other.0.len() == 1) {
+            (true, true) => {
+                self.0[0].compare_single_path_options_complexity_out_of_context(&other.0[0])
+            }
+            (true, false) => {
+                self.0[0].compare_single_vs_multi_path_options_complexity_out_of_context(other)
+            }
+            (false, true) => Ok(other.0[0]
+                .compare_single_vs_multi_path_options_complexity_out_of_context(self)?
+                .reverse()),
+            (false, false) => Ok(Ordering::Equal),
+        }
```

Comment:
Slice patterns can extract the item based on the length of the slice value:
```suggestion
        match (self.0.as_slice(), other.0.as_slice()) {
            ([a], [b]) => {
                a.compare_single_path_options_complexity_out_of_context(b)
            }
            ([a], _) => {
                a.compare_single_vs_multi_path_options_complexity_out_of_context(other)
            }
            (_, [b]) => Ok(b
                .compare_single_vs_multi_path_options_complexity_out_of_context(self)?
                .reverse()),
            _ => Ok(Ordering::Equal),
        }
```